### ISSUE TITLE
Performance gap on NERF synthetic dataset

### ISSUE THREAD

**Issue by walsvid** [yashbhalgat/HashNeRF-pytorch#4]
As mentioned in another issue #2, using the default config like nerf-pytorch does not get comparable performance to Instant-NGP.

**Comment by yashbhalgat** [comment]
Hi @walsvid, this is indeed the case, thanks for pointing out! Although the renders look "good", the PSNR is not as good as reported values in Instant-NGP.

**Note:** You should look at the _testset_ PSNR (not the training PSNR). I have just pushed some code to print these values. Pull the latest `master` branch.

Here's an example with Lego dataset:
PSNR on test set = `28.89`. Corresponding renders:

https://user-images.githubusercontent.com/8559512/154839391-afd26318-b6b7-4737-b5e2-091cc804bce5.mp4

**Comment by ashawkey** [comment]
A possible reason is that instant-ngp loads all train/val/test data to train (see [here](https://github.com/NVlabs/instant-ngp/blob/7b37eee0aa9ea31362d2a5a799e007a731de385c/src/testbed_nerf.cu#L2211)) for nerf_synthetic dataset?

**Comment by Xuanqing-C** [comment]
Seems that if we add the TV_loss's iteration, PSNR could rise. (in run_nerf.py line 876)

**Comment by Feynman1999** [comment]
so, can anyone give a benchmark for this repo v.s. official instant-ngp repo? it would be very useful!

**Comment by shreyk25** [comment]
Hi @yashbhalgat , may I know what is the expected difference in training speed of This repo, pytorch-nerf and instant-ngp?
I ran the pytorch-nerf and this repo and did not find any such significant gain in speed (which is the main purpose of instant-ngp)
Thank You.

**Comment by yashbhalgat** [comment]
@Feynman1999 regarding benchmarks v/s Instant-NGP, I will try to get this ready, but it won't be any time soon as I am currently caught up with other projects. If you or someone else can work on this, feel free to open a pull request. :) 

@shreyk25, 
1. this is a pure PyTorch(+Python) implementation, so it isn't as fast as the CUDA(C++) implementation by [Instant-NGP](https://github.com/NVlabs/instant-ngp). 
2. Compared to [nerf-pytorch](https://github.com/yenchenlin/nerf-pytorch), the **iterations/second** speed during training is almost the same. Although, as mentioned in the README, the HashNeRF algorithm converges much faster (refer the video below). For the "chair" dataset, HashNeRF converges in about 3000 iterations (which translates to 15 minutes training time) while vanilla NeRF ([nerf-pytorch](https://github.com/yenchenlin/nerf-pytorch)) will take a few hours to reach similar performance. I have observed approximately 20x _convergence_ speedup compared to [nerf-pytorch](https://github.com/yenchenlin/nerf-pytorch).

https://user-images.githubusercontent.com/8559512/154065666-f2eb156c-333c-4de4-99aa-8aa15a9254de.mp4

Hope this helps. :)

**Comment by Feynman1999** [comment]
> Seems that if we add the TV_loss's iteration, PSNR could rise. (in run_nerf.py line 876)

have you tried this tvloss? how it affect psnr?

**Comment by leo-frank** [comment]
I verify that "hash encoding + small MLP" has faster convergence speed than “vanilla large MLP”, see this figure (Bottom is HashNeRF, while the above is vanilla NeRF) :

at 500 iteration:
![image](https://github.com/yashbhalgat/HashNeRF-pytorch/assets/47302141/d0604064-81a8-49ce-809e-fe9e8c23abc1)
at 1000 iteration:
![image](https://github.com/yashbhalgat/HashNeRF-pytorch/assets/47302141/30c2cd46-9290-4807-b4a9-57d7218eff8b)

**Comment by zParquet** [comment]
Thanks for the great work @yashbhalgat , do you now have any idea about the performance (numerical results) gap between instant and this implementation? I guess there may be some lack of essential implementation details in this repo, but I am not able to find it out...

**Comment by MrMois** [comment]
@zParquet I am wondering the same and searching for quite a while now, because the gap is quite large. Two differences I found so far are:

1) Here, small eps for Adam is only applied to hash table entries, yet in the paper it seems they use it also for the MLP.
https://github.com/yashbhalgat/HashNeRF-pytorch/blob/425e70bab401c4f0dca61dbbc7b9518035b7e38e/run_nerf.py#L265

2) They mention the mapping between grid and table to be 1:1 on lower resolutions. This injectivity is not given in this implementation, because in theory there could also happen collisions on the lower resolutions. However they are very unlikely, which makes the effect on PSNR questionable.

Further ideas are welcomed :D

This work seems to reach atleast PSNR 34, however they are using an own Cuda version of the encoding.
https://github.com/ashawkey/torch-ngp

**Comment by MrMois** [comment]
After reading the appendix E.3, it seems a huge benefit comes from a (very) large number of rays in each batch, at the cost of fewer samples. However these fewer samples seem to be possible because of their additional nested occupancy grids.

![image](https://github.com/yashbhalgat/HashNeRF-pytorch/assets/43606051/ae60e55c-ff9d-428e-bea2-aff4b0c4c36c)
